### ISSUE TITLE
apply_dimension: too UDF centric

### ISSUE THREAD

**Issue by soxofaan** [Open-EO/openeo-python-client#137]
current implementations of `apply_dimension` are hardcoded for UDF's, but they should also allow non-UDF usage (e.g see https://openeo.org/documentation/1.0/processes.html#apply_dimension)

https://github.com/Open-EO/openeo-python-client/blob/d2b80160512d9aff90f6a7552f5dd3d528bfe20f/openeo/rest/imagecollectionclient.py#L530

https://github.com/Open-EO/openeo-python-client/blob/d2b80160512d9aff90f6a7552f5dd3d528bfe20f/openeo/rest/datacube.py#L422

**Comment by soxofaan** [comment]
Sorry I was too fast: non-udf usage is supported (but only for single node "callbacks")
Nevertheless `apply_dimension` is still too coupled with UDF's usage, does not align with comparable processes like `reduce_dimension` and should also support non-single process callbacks

**Comment by jdries** [comment]
I believe this is solved? Current signature is:
```def apply_dimension(
            self, code: str = None, runtime=None,
            process: [str, PGNode, typing.Callable] = None,
            version="latest", dimension='t', target_dimension=None
    ) -> 'DataCube':
```
So complex callbacks are possible. More improvements would require incompatible changes.

**Comment by soxofaan** [comment]
complex callbacks are indeed possible now.

the remaining "problem" is that `apply_dimension` still has these arguments `code`, `runtime` and `version`, which are UDF-specific. It would be cleaner (boilerplate-wise) to abstract/encapsulate this in the the `process` argument in some way. 
But indeed, that would be backward-incompatible change. 

A solution would be to replace `code`, `runtime` and `version`  with `**kwargs`  to still support old-style calls (with deprecation warning), but without polluting the argument list too much

**Comment by soxofaan** [comment]
As raised in openEOPlatform/architecture-docs#144 (and related devtelco), the current mixed nature of `apply_dimension` leads to confusion.

Brainstorming a bit here. Current signature is:
```python
    def apply_dimension(self,
            code: str = None,
            runtime=None,
            process: [str, PGNode, typing.Callable] = None,
            version="latest", 
            dimension='t', 
            target_dimension=None
    )
```

Ideally (for UX and internal consistency) we go to something that is closer to the official spec, e.g.:
```python
    def apply_dimension(self,
            process: [str, PGNode, typing.Callable] = None,
            dimension='t', 
            target_dimension=None
    )
```
Which we can't switch to right away without risking compatibility issues. However, with a bit of argument sniffing and kwargs tricks it's possible to get close to that while still offering backwards compatibility.

Let's assume these are the (old style and new style) calls we want to keep working:
```python
# Old style UDF-centric calls with at least a source code string `code` (as positional args or with kwargs)
cube.apply_dimension(code)  
cube.apply_dimension(code, runtime)
cube.apply_dimension(code, runtime, dimension=dimension)
cube.apply_dimension(code=code)
cube.apply_dimension(code=code, runtime=runtime) 
...

# Non-UDF-centric calls currently only make sense with keyword arguments
cube.apply_dimension(process=process)
cube.apply_dimension(process=process, dimension="t")
```

I think it's possible to support this with something like this:
```python
    def apply_dimension(self,
            process: [str, PGNode, typing.Callable] = None,
            dimension='t', 
            target_dimension=None,
            **kwargs
    ):
        if (isinstance(process, str) and "\n" in process) or "code" in kwargs:
            # Detected old style UDF-centric argument call: recover UDF-style arguments 
            code = kwargs.get("code", process)
            runtime = kwargs.get("runtime") or dimension
            if runtime in ("t", "temporal"):
                runtime = None # (legacy default of `runtime`)
            version = kwargs.get("version", "latest")
            # Convert UDF arguments to a new style "UDF "process
            process = self._create_ run_udf(code, runtime, version)
     
     # Continue with `process`
```

**Comment by soxofaan** [comment]
This is related to #312  and being addressed in #313

**Comment by soxofaan** [comment]
Just merged #313   which addressed this issue: old style `apply_dimension` calls are still supported, but will raise deprecation warnings, while new style calls are promoted (e.g. in docs).

Next step (for later) is to also remove support for old patterns: #330 

I think we can close this ticket now
